### PR TITLE
first round of face related api fixes

### PR DESCRIPTION
**Pull request by headupinclouds**
* drishti/FaceTracker.hpp updates
   + fix index and resize output arrays
   + make drishti::sdk::Face a lightweight struct (eyes + landmarks) w/ public Array<> 
   + add public <-> private face model conversion for Face
* don’t wait on scene in FaceFinder::~FaceFinder() destructor if optimized pipeline is not used
* core/convert.cpp: fix bounds access in unpack()
* move private headers out of DRISHTI_DRISHTI_HDRS_PUBLIC
* disable experimental flow in public FaceTracker API
* Add a public SDK wrapper for SensorModel class
* disable shared lib by default for hunter packaging
* DRISHTI_BUILD_EXAMPLES=ON
* For default repository builds it is desirable to build the console apps and examples with DRISHTI_BUILD_EXAMPLES=ON — we can disable these explicitly for hunter lib only package use.
 * add DRISHTI_INSTALL option, skip installation stage if off (default on)
* wrap add_subdirectory(app) w/ DRISHTI_BUILD_EXAMPLES
* set DRISHTI_BUILD_EXAMPLES=ON (be explicit) for CI builds
* remove legacy DRISHTI_USE_TEXT_ARCHIVES from build-appveyor.cmd
* remove “.” from option descriptions
* DRISHTI_BUILD_{TESTS,BENCHMARKS}=OFF by default
* Enable tests in CI configs explicitly
* update packages in Config.cmake.in (Eigen -> Eigen3)
* add public API layer for Sensor model
* add a handful of getters/setters for FaceTracker configurations
* add a Matrix<T,int,int> API type

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -386,15 +386,16 @@ option(BUILD_SHARED_LIBS "Build package with shared libraries." OFF)
 #################################
 
 # This will build a static library using all specified modules
-option(DRISHTI_BUILD_WORLD "Build single library" ON)
+option(DRISHTI_INSTALL "Run package-config installation" ON)
```
Comment (ruslo) [inline CMakeLists.txt]: What is the point of this option? If user don't want to install this library then he will not call `cmake --build _builds --target install`. Is this for 3rd parties when Drishti added by `add_subdirectory(3rdParty/Drishti)`?

Comment (headupinclouds) [reply CMakeLists.txt]: Yes, that was the intention.  I'm not currently testing in this mode, but I've added such flags to other libs in the past for this reason.   Probably inverting this option (i.e., `DRISHTI_SKIP_INSTALL`) is a little clearer for such a use case.  We can rename or remove.

**Comment by headupinclouds** [comment]
There seems to be some gcc + lto strangeness.  Maybe this:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53808

**Comment by ruslo** [comment]
> https://stackoverflow.com/questions/25878407/how-can-i-use-lto-with-static-libraries

Should not be the reason. `gcc-ranlib` and `gcc-ar` set by CMake.

> https://stackoverflow.com/questions/25878407/how-can-i-use-lto-with-static-libraries

Looks like this one is fixed in GCC 4.9+

**Comment by headupinclouds** [comment]
I do notice these `BFD  ... plugin needed to handle lto object` warnings when building the static libs.  My sense is that this issue is related to the untested LTO + static configuration.  A small tests case is needed.  I might have to take a look this weekend.  

```
BFD: /Users/dhirvonen/devel/drishti_hunter_test/src/3rdparty/drishti/_builds/android-ndk-r10e-api-19-armeabi-v7a-neon-hid-sections-lto-MinSizeRel/src/lib/drishti/st13drPy/Eye.cpp.o: plugin needed to handle lto object
BFD: /Users/dhirvonen/devel/drishti_hunter_test/src/3rdparty/drishti/_builds/android-ndk-r10e-api-19-armeabi-v7a-neon-hid-sections-lto-MinSizeRel/src/lib/drishti/st13drPy/EyeIO.cpp.o: plugin needed to handle lto object
```

While investigating this, I was reminded that the surrogate internal-static-library-as-replacement-for-object-library structure used for the versioned shared public api layer doesn't feel right for the final static build configuration.  I guess if hunter is used, the additional internal private library layer should be fairly transparent.  We could either rebuild the files needed in the public api using sources collect by `sugar` or we could potentially use the static merge feature.  I think it needs more thought.

**Comment by headupinclouds** [comment]
```
[ 82%] Linking CXX executable test-drishti-drishti-face

cd /Users/travis/build/elucideye/drishti/_builds/android-ndk-r10e-api-19-armeabi-v7a-neon-hid-sections-lto-MinSizeRel/src/lib/drishti/drishti/ut && /Users/travis/build/elucideye/drishti/_ci/cmake/bin/cmake -E cmake_link_script CMakeFiles/test-drishti-drishti-face.dir/link.txt --verbose=1


/Users/travis/build/elucideye/drishti/_ci/android-ndk-r10e/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-g++ --sysroot=/Users/travis/build/elucideye/drishti/_ci/android-ndk-r10e/platforms/android-19/arch-arm  -fvisibility-inlines-hidden -fvisibility=hidden -fdata-sections -ffunction-sections -std=c++11   -march=armv7-a -marm -mfpu=neon -mfloat-abi=softfp -funwind-tables -no-canonical-prefixes -fexceptions -frtti -Wno-c++11-narrowing -Wno-unused-parameter -Wno-unused-local-typedefs -Wno-unknown-attributes -Wno-shorten-64-to-32 -fopenmp -Os -DNDEBUG -flto -fno-fat-lto-objects  -Wl,--fix-cortex-a8  -fPIE -pie -Wl,--gc-sections -Wl,-z,nocopyreloc  CMakeFiles/test-drishti-drishti-face.dir/test-FaceTracker.cpp.o  -o test-drishti-drishti-face ../../libdrishti-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/sdk/native/libs/armeabi-v7a/libopencv_world-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libgtest-MinSizeRel.a ../../libdrishtisdk-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libaglet-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libgauze-MinSizeRel.a ../../libdrishti_world-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libcvmatio-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/sdk/native/libs/armeabi-v7a/libopencv_world-MinSizeRel.a -ldl -llog /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libpng-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libz-MinSizeRel.a /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libxgboost-MinSizeRel.a -lm /Users/travis/.hunter/_Base/4b72d50/5449f57/0e66ba9/Install/lib/libogles_gpgpu-MinSizeRel.a ../../../../../../../_ci/android-ndk-r10e/platforms/android-19/arch-arm/usr/lib/liblog.so ../../../../../../../_ci/android-ndk-r10e/platforms/android-19/arch-arm/usr/lib/libandroid.so ../../../../../../../_ci/android-ndk-r10e/platforms/android-19/arch-arm/usr/lib/libEGL.so ../../../../../../../_ci/android-ndk-r10e/platforms/android-19/arch-arm/usr/lib/libGLESv2.so  "/Users/travis/build/elucideye/drishti/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libgnustl_static.a"


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/iostream:62:0: warning: type of 'cerr' does not match original declaration


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/iostream:62:18: warning: type of 'cerr' does not match original declaration


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/iostream:62:0: note: previously declared here


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/sstream:387:11: warning: type 'struct basic_ostringstream' violates one definition rule


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/sstream:387:11: note: a type with the same name but different layout is defined in another translation unit


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/ostream:58:11: warning: type 'struct basic_ostream' violates one definition rule


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/ostream:58:0: note: a type with the same name but different layout is defined in another translation unit


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/bits/basic_ios.h:66:11: warning: type 'struct basic_ios' violates one definition rule


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/bits/basic_ios.h:66:11: note: a type with the same name but different layout is defined in another translation unit


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/istream:58:11: warning: type 'struct basic_istream' violates one definition rule


/Users/travis/build/elucideye/drishti-upload/_ci/android-ndk-r10e/sources/cxx-stl/gnu-libstdc++/4.9/include/istream:58:11: note: a type with the same name but different layout is defined in another translation unit


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans0.ltrans.o:ccRRA3ox.ltrans0.o:function (anonymous namespace)::FaceTest_RunSimpleTest_Test::TestBody() [clone .lto_priv.993]: error: undefined reference to 'drishti::sdk::FaceTracker::good() const'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans0.ltrans.o:ccRRA3ox.ltrans0.o:function (anonymous namespace)::FaceTest_RunSimpleTest_Test::TestBody() [clone .lto_priv.993]: error: undefined reference to 'drishti::sdk::FaceTracker::operator()(drishti::sdk::VideoFrame const&)'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function std::_Sp_counted_ptr_inplace<drishti::sdk::FaceTracker, std::allocator<drishti::sdk::FaceTracker>, (__gnu_cxx::_Lock_policy)2>::_M_dispose(): error: undefined reference to 'drishti::sdk::FaceTracker::~FaceTracker()'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::SensorModel::Intrinsic::Intrinsic(drishti::sdk::Vec<float, 2> const&, float, drishti::sdk::Vec<int, 2> const&)'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::SensorModel::Extrinsic::Extrinsic(drishti::sdk::Matrix<float, 3, 3> const&)'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::SensorModel::SensorModel(drishti::sdk::SensorModel::Intrinsic const&, drishti::sdk::SensorModel::Extrinsic const&)'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::Context::Context(drishti::sdk::SensorModel&)'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::FaceTracker::FaceTracker(drishti::sdk::Context*, drishti::sdk::FaceTracker::Resources&)'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::Context::~Context()'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::SensorModel::~SensorModel()'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::Context::~Context()'


/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T//ccRRA3ox.ltrans1.ltrans.o:ccRRA3ox.ltrans1.o:function (anonymous namespace)::FaceTest::create(cv::Size_<int> const&, int, bool) [clone .isra.276] [clone .lto_priv.1050]: error: undefined reference to 'drishti::sdk::SensorModel::~SensorModel()'
```

**Comment by headupinclouds** [comment]
https://gcc.gnu.org/wiki/LinkTimeOptimizationFAQ

Creating an archive with gcc-ar works as expected:
 
```
 gcc -c -flto a.c
 gcc-ar cr a.a a.o
 gcc -flto b.c a.a
```
 
If you use a version of ar that doesn't load the liblto_plugin, you'll get linker errors:

```
 gcc -c -flto a.c
 ar cr a.a a.o
 gcc -flto b.c a.a
```

```
/tmp/ccncHnsT.ltrans0.ltrans.o:ccncHnsT.ltrans0.o:function foo1: error: undefined reference to 'foo3'
/tmp/ccncHnsT.ltrans0.ltrans.o:ccncHnsT.ltrans0.o:function main: error: undefined reference to 'foo2'
collect2: error: ld returned 1 exit status
```

**Comment by ruslo** [comment]
I think we've hit this issue before, no? Something about inappropriate usage of `ranlib`. If we are using Hunter to manage Drishti library we don't need "merge-everything-into-big-static-library" step. All dependent libraries will be linked automatically, user can just use `Drishti::drishti` or something.

**Comment by headupinclouds** [comment]
> I think we've hit this issue before, no? Something about inappropriate usage of ranlib

That sound familiar.  Another [lto SO issue](https://stackoverflow.com/a/39256013) 

Here are the current CMake tools  w/ the active Android toolchain.
```
CMAKE_RANLIB /android/android-ndk-r10e/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-ranlib
CMAKE_AR /android/android-ndk-r10e/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-ar
CMAKE_NM /android/android-ndk-r10e/toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-nm
CMAKE_CXX_ARCHIVE_CREATE <CMAKE_AR> qc <TARGET> <LINK_FLAGS> <OBJECTS> 
CMAKE_CXX_ARCHIVE_FINISH <CMAKE_RANLIB> <TARGET>
```

> Should not be the reason. gcc-ranlib and gcc-ar set by CMake.

That could be the issue.  I will take a look.

File: src/lib/drishti/CMakeLists.txt [hunk 2]
```diff
@@ -373,8 +373,10 @@ else()
   drishti_strip(drishti)
 endif()
 
-drishti_split_debug_symbols(drishti)
-
+if(DRISHTI_BUILD_SHARED_SDK)  
+  drishti_split_debug_symbols(drishti)
```
Comment (headupinclouds) [inline src/lib/drishti/CMakeLists.txt]: This was the culprit

**Comment by headupinclouds** [comment]
It turned out all of the `lto` problems were really due to a `strip` as part of creating resymbolicification friendly releases with shared library builds.

https://github.com/elucideye/drishti/pull/558/commits/1e389f983924e1d302433c006e170dbff952ef55#r136215175

**Comment by headupinclouds** [comment]
> All dependent libraries will be linked automatically, user can just use Drishti::drishti or something.

Yes, but I think the drishtisdk layer could be removed and isn't really needed with drishti_world (which could be renamed).  This was originally a target for fusing object libs: `drishti->drishtisdk->drishti_{world|eye,face,*}`.

**Pull request merged**